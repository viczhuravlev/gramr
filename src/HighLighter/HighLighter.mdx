---
name: HighLighter
---

import { Playground, PropsTable } from 'docz';
import HighLighter from './HighLighter';

# HighLighter

<PropsTable of={HighLighter} />

## Basic usage

<Playground>
  {() => {
    const search = 'ui';
    const text =
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';
    return <HighLighter text={text} search={search} />;
  }}
</Playground>

## Custom color

<Playground>
  {() => {
    const search = 'qui';
    const text =
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';
    return <HighLighter text={text} search={search} color="skyBlue" opacity={1} />;
  }}
</Playground>

## Not found

<Playground>
  {() => {
    const search = 'asdasdasdas';
    const text =
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';
    return <HighLighter text={text} search={search} color="skyBlue" opacity={1} />;
  }}
</Playground>

## Empty search

<Playground>
  {() => {
    const search = null;
    const text =
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';
    return <HighLighter text={text} search={search} color="skyBlue" opacity={1} />;
  }}
</Playground>

## Empty text

<Playground>
  {() => {
    const search = 'ui';
    const text = null;
    return <HighLighter text={text} search={search} color="skyBlue" opacity={1} />;
  }}
</Playground>
