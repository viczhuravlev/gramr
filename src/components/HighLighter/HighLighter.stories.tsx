import React from 'react';
import { storiesOf } from '@storybook/react';

import { HighLighter } from './index';

storiesOf('Typography/HighLighter', module)
  .add('Default', () => (
    <HighLighter search="consectetur adipiscing elit">
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
      tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
      veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea
      commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
      velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat
      cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id
      est laborum.
    </HighLighter>
  ))
  .add('Custom color', () => (
    <HighLighter search="ui" color="#1890ff" opacity={0.5}>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
      tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
      veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea
      commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
      velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat
      cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id
      est laborum.
    </HighLighter>
  ))
  .add('Not found', () => (
    <HighLighter search="asdasd">
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
      tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
      veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea
      commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
      velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat
      cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id
      est laborum.
    </HighLighter>
  ))
  .add('Empty search', () => (
    <HighLighter search={null} color="#1890ff" opacity={1}>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
      tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
      veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea
      commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
      velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat
      cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id
      est laborum.
    </HighLighter>
  ));
