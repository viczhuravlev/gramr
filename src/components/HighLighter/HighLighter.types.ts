import { CSSProperties } from 'react';

export interface HighLighterDefaultProps {
  color: string;
  opacity: number;
}

export interface HighLighterProps extends Partial<HighLighterDefaultProps> {
  style?: CSSProperties;
  search?: string | null;
  children: string;
  className?: string;
}
