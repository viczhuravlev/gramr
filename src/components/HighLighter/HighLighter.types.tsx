export interface HighLighter {
  /**
   * Default:  0.1
   */
  opacity?: number;
  /**
   * Default: #F7981C
   */
  color?: string;
}

export interface HighLighterProps extends HighLighter {
  text?: string;
  search?: string;
}
