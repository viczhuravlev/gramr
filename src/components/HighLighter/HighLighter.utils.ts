export function searchText(text: string, search: string): string[] {
  const pattern = new RegExp(`(${search})`, 'i');

  return text.split(pattern);
}
