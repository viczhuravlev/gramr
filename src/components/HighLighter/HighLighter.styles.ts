import { styled } from 'linaria/react';

import { hexToRgb } from '../../utils/styles';

import * as T from './HighLighter.types';

export const HighLighter = styled.span<T.HighLighter>`
  background-color: ${(props) =>
    hexToRgb(props.color || '#F7981C', props.opacity) || 'transparent'};
`;
