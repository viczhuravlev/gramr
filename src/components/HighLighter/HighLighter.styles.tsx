import styled from '@emotion/styled';

import * as T from './HighLighter.types';

export const HighLighter = styled.span<T.HighLighter>`
  background-color: ${props => props.color};
  opacity: ${props => props.opacity};
`;
