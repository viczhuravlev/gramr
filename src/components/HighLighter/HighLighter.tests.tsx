import React from 'react';
import { mount } from 'enzyme';

import HighLighter from './HighLighter';
// import * as S from './HighLighter.styles';
// import * as T from './HighLighter.types';

// const styleColor = 'skyBlue';
// const styleOpacity = 12;
const search = 'ui';
const text =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';

describe('[component] HighLighter', () => {
  it('[Snapshot] HighLighter', () => {
    const component = mount(<HighLighter text={text} search={search} />);

    expect(component).toMatchSnapshot();
  });

  // it('[props] search is empty', () => {
  //   const component = shallow(<HighLighter text={text} />);
  //
  //   expect(component.text()).toEqual(text);
  // });
  //
  // it('[props] search is wrong', () => {
  //   const searchWrong = 'asdasdasdas';
  //   const component = shallow(<HighLighter text={text} search={searchWrong} />);
  //
  //   expect(component.text()).toEqual(text);
  // });
  //
  // it('[props] text is empty', () => {
  //   const component = shallow(<HighLighter search={search} />);
  //
  //   expect(component.html()).toEqual('');
  // });
  //
  // it('[props] text is not string', () => {
  //   // @ts-ignore
  //   const component = shallow(<HighLighter search={styleOpacity} />);
  //
  //   expect(component.html()).toEqual('');
  // });
  //
  // it('[props] Color is custom ', () => {
  //   const component = mount<T.HighLighterProps>(
  //     <HighLighter text={text} search={search} color={styleColor} opacity={styleOpacity} />
  //   );
  //
  //   const styleComponents = component.find<T.HighLighter>(S.HighLighter);
  //
  //   // @ts-ignore
  //   [...styleComponents].forEach(item => {
  //     const { color, opacity, children } = mount<{ children: ReactNode } & T.HighLighter>(
  //       item
  //     ).props();
  //
  //     expect(children).toEqual(search);
  //     expect(color).toEqual(styleColor);
  //     expect(opacity).toEqual(styleOpacity);
  //   });
  // });
});
