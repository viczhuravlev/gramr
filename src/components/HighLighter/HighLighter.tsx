import React from 'react';

import * as T from './HighLighter.types';
import * as U from './HighLighter.utils';
import * as S from './HighLighter.styles';
import { defaultLight } from '../../styles/theme';

function HighLighter(props: T.HighLighterProps): JSX.Element | null {
  const { style, color, search, opacity, children, className } = props;

  if (!children) {
    return null;
  }

  if (!search) {
    return <>{children}</>;
  }

  const resultSearch = U.searchText(children, search);
  const searchLower = search.toLowerCase();

  if (resultSearch?.length < 2) {
    return <>{children}</>;
  }

  return (
    <>
      {resultSearch.map((word, index) => {
        if (word.toLowerCase() !== searchLower) {
          return word;
        }

        return (
          <S.HighLighter
            key={index}
            color={color || defaultLight['--gramr-ui-prominent-color']}
            style={style}
            opacity={opacity || 0.4}
            className={className}
          >
            {word}
          </S.HighLighter>
        );
      })}
    </>
  );
}

const defaultProps: T.HighLighterDefaultProps = {
  color: defaultLight['--gramr-ui-prominent-color'],
  opacity: 0.4,
};

HighLighter.defaultProps = defaultProps;

export default HighLighter;
