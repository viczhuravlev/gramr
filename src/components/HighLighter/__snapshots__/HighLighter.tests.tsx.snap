// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[component] HighLighter [Snapshot] HighLighter 1`] = `
<HighLighter
  color="#F7981C"
  opacity={0.2}
  search="ui"
  text="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, q
  <Styled(span)
    color="#F7981C"
    key="1"
    opacity={0.2}
  >
    <span
      className="css-ecly6w"
      color="#F7981C"
      opacity={0.2}
    >
      ui
    </span>
  </Styled(span)>
  s nostrud exercitation ullamco laboris nisi ut aliq
  <Styled(span)
    color="#F7981C"
    key="3"
    opacity={0.2}
  >
    <span
      className="css-ecly6w"
      color="#F7981C"
      opacity={0.2}
    >
      ui
    </span>
  </Styled(span)>
  p ex ea commodo consequat. D
  <Styled(span)
    color="#F7981C"
    key="5"
    opacity={0.2}
  >
    <span
      className="css-ecly6w"
      color="#F7981C"
      opacity={0.2}
    >
      ui
    </span>
  </Styled(span)>
  s aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa q
  <Styled(span)
    color="#F7981C"
    key="7"
    opacity={0.2}
  >
    <span
      className="css-ecly6w"
      color="#F7981C"
      opacity={0.2}
    >
      ui
    </span>
  </Styled(span)>
   officia deserunt mollit anim id est laborum.
</HighLighter>
`;
